### DISCUSSION TITLE
ObservableObject Inspection Abnormalities?

### DISCUSSION THREAD

**Discussion by perlguy99** [nalexn/ViewInspector#328] [Q&A]
I created a simple Tic-Tac-Toe game to help learn ViewInspector better...

Repo: [Public Repo](https://github.com/perlguy99/TicTacToeViewInspector)

I've got a GameController that is an `ObservableObject` has 2 `@Published` properties, `gameResult` and `currentTurn`, the GameController is also responsible for the `gameHeaderTitle` String...

```swift
class GameController: ObservableObject {
    @Published var gameResult: GameResult = .inProgress
    @Published var currentTurn: Turn = .x
    
    var gameBoard: [Square] = []
    
    var gameHeaderTitle: String {
        switch gameResult {
        case .inProgress:
            return currentTurn.turnString
        case .winner(let winner):
            return "Winner: \(winner == .x ? "X" : "O")"
        case .draw:
            return "DRAW"
        }
    }

... rest of code
```

In my GameBoardView I have 
```swift
@ObservedObject var controller = GameController()
```

### On to the ViewInspector Question
I have the following 2 tests in my code, I stripped them down to just what is needed for this issue...

1 test uses the way described for `@ObservedObjects` and the other uses the way to tests things like `@State` values...

```swift
    func testGameBoardTitleIsCorrect_Q1a() throws {
        // Given that I have a Game Board
        let sut = GameBoardView()
        
        let displayedText = { try sut.inspect().find(viewWithTag: "GameBoardViewTitle").text().string() }
        
        XCTAssertEqual(try displayedText(), "It's X's turn")
        sut.controller.takeTurnAt(0) // X
        XCTAssertEqual(try displayedText(), "It's O's turn")
    }

    func testGameBoardTitleIsCorrect_MoreViewInspector_Q1b() throws {
        // Given that I have a Game Board
        var sut = GameBoardView()
        
        let exp = sut.on(\.didAppear) { view in
            XCTAssert(try view.actualView().inspect().find(viewWithTag: "GameBoardViewTitle").text().string() == "It's X's turn")
            try view.actualView().controller.takeTurnAt(0)
            XCTAssert(try view.actualView().inspect().find(viewWithTag: "GameBoardViewTitle").text().string() == "It's O's turn")
        }
        
        ViewHosting.host(view: sut)
        wait(for: [exp], timeout: 0.1)
    }

```

Those tests pass just fine and the app works, however...

If I remove the `@Published` from the `currentTurn` variable, ALL of the test _still_ pass, but the `gameHeaderTitle` doesn't update the turns when playing the game!

`gameHeaderTitle` is not `@Published`, yet when there is a "winner", then it updates properly?!

Did I setup the tests wrong, or is there something wrong when inspecting the View?

I'm a bit confused at the moment, anyone have any ideas?

**Comment by Zelzazor** [comment]
`ObservableObjects` are, first of all, classes. You can update the public properties of any class, and then assert the resulting value of said update. 

The `ObservableObject` conformity, and most important, the use of `@Published` does not alter that behavior. In the context of a subscriber of an ObservableObject, like a `View`, the behavior changes. Here, any update to properties annotated with `@Published` triggers the method `objectWillChange`, and said subscriber use that to see if it should execute something (in the case of Views, if it should trigger a rerender)

Hope that clarifies it!

Edit: tagged someone by mistake, was writing this on mobile
